File: src/renderer/vulkan/FrameSync.cpp
////////////////////////////////////////////////////////////////////////////////
/// Reaper
///
/// Copyright (c) 2015-2022 Thibault Schueller
/// This file is distributed under the MIT License
////////////////////////////////////////////////////////////////////////////////

#include "FrameSync.h"

#include "Backend.h"
#include "Debug.h"
#include "api/AssertHelper.h"

namespace Reaper
{
FrameSyncResources create_frame_sync_resources(VulkanBackend& backend)
{
    const VkSemaphoreTypeCreateInfo semaphore_type_create_info = {
        .sType = VK_STRUCTURE_TYPE_SEMAPHORE_TYPE_CREATE_INFO_KHR,
        .pNext = nullptr,
        .semaphoreType = VK_SEMAPHORE_TYPE_TIMELINE_KHR,
        .initialValue = 0,
    };

    const VkSemaphoreCreateInfo timeline_semaphore_create_info = {
        .sType = VK_STRUCTURE_TYPE_SEMAPHORE_CREATE_INFO,
        .pNext = &semaphore_type_create_info,
        .flags = VK_FLAGS_NONE,
    };

    VkSemaphore timeline_semaphore;
    AssertVk(vkCreateSemaphore(backend.device, &timeline_semaphore_create_info, nullptr, &timeline_semaphore));

    VulkanSetDebugName(backend.device, timeline_semaphore, "Main frame sync timeline semaphore");

    FrameSyncResources resources = {};

    resources.timeline_semaphore = timeline_semaphore;

    return resources;
}

void destroy_frame_sync_resources(VulkanBackend& backend, const FrameSyncResources& resources)
{
    vkDestroySemaphore(backend.device, resources.timeline_semaphore, nullptr);
}
} // namespace Reaper

File: src/renderer/vulkan/frame_sync.zig
// Port of src/renderer/vulkan/FrameSync.h + FrameSync.cpp
//
// A single timeline semaphore paces the CPU against the GPU: frame N submits
// with a signal value of N, and the next frame waits on N before recording.

const std = @import("std");
const vk = @import("vulkan");

pub const FrameSyncResources = struct {
    timeline_semaphore: vk.Semaphore,
};

pub fn create(vkd: anytype, device: vk.Device) !FrameSyncResources {
    const semaphore_type_create_info = vk.SemaphoreTypeCreateInfo{
        .s_type = .semaphore_type_create_info,
        .p_next = null,
        .semaphore_type = .timeline,
        .initial_value = 0,
    };

    const timeline_semaphore_create_info = vk.SemaphoreCreateInfo{
        .s_type = .semaphore_create_info,
        .p_next = @ptrCast(&semaphore_type_create_info),
        .flags = .{},
    };

    return .{
        .timeline_semaphore = try vkd.createSemaphore(device, &timeline_semaphore_create_info, null),
    };
}

pub fn destroy(vkd: anytype, device: vk.Device, resources: FrameSyncResources) void {
    vkd.destroySemaphore(device, resources.timeline_semaphore, null);
}
